wordpress-mobile/WordPress-Android
# Apply user language selected in WordPress app to Jetpack app during migration

Fixes #17685

This PR enhances the implementation of the Jetpack Content Migration Flow by applying the interface language set by the user in the WordPress app to the Migration screens.

> **Note**:
> During implementation I experimented with the widely popular approach of inheriting from `LocaleAwareActivity` in the `JetpackMigrationActivity`, but it did not work correctly in cases when the WordPress app theme is also customized by the user:
> ℹ️ More specifically upon navigating to another non-migration specific screen (Help / My Site / Login), if the user changes the system theme to be different than the custom app theme migrated from the WordPress app (system theme = light, Jetpack app theme = dark) the next screen could end up not translated into the user language.
> ℹ️ On top of this issue, using `LocaleAwareActivity` is not working correctly with `Compose`, or at least not for the `JetpackMigrationActivity`, where recreating the fragment is avoided in order fix the [error when changing the device settings](https://github.com/wordpress-mobile/WordPress-Android/pull/17659) mid-flow.
> 🚀 **As a solution** to these issues I introduced the new `LocaleAwareComposable` utility that listens to language changes and synchronizes the local configuration inside its `content` composable when needed.


To test:
1. Build, install, launch the WordPress app and login
2. Go to `Me` → `App Settings` → `Appearance` and Set the WordPress app theme to `Dark` mode (or the opposite of the system default)
2. Go to `Interface Language` and change the app language to `Arabic` (or a language you know, `Arabic` helps to verify the RTL texts)
3. Build and install the Jetpack app of the same flavor
4. Open Jetpack app and **expect** that the migration Welcome screen is shown
5. **Verify** on the migration welcome screen:
   - The `app theme` is in `Dark` mode (or whatever was manually set in WordPress)
   - The `language` is `Arabic` (or whatever was manually set in WordPress)
   - The `layout direction` is `RTL ` mode if Arabic/ `LTR` if in a LTR language
6. Explore the different options to interrupt & resume the flow, eg:
   - Tap back to end up on Login Screen, force close & restart the app
   - Tap `Need help?` → `Log out`, force close & restart the app
   - Complete the migration flow by tapping on the primary green button
     - On `My Site` tap on the card with the `WordPress` logo (to open "Please Delete WordPress.." screen)
   - Eventually set the system theme to `dark` mode, then to `light mode` and navigate to another screen (there was an issue with this behavior and specific fixes were needed for this)
 7. **Verify** the `app theme`, `language` and `layout direction` are persisted (same as on Welcome Screen)

<details>
<summary><h4>🎬 Video recording preview</h4> (click to expand)</summary>

https://user-images.githubusercontent.com/4588074/212742532-c8f76ae2-d86a-4918-947b-889c5b77b2cf.mp4
</details>

## Regression Notes
1. Potential unintended areas of impact
   Migration of user language and theme from WordPress to Jetpack app.

5. What I did to test those areas of impact (or what existing automated tests I relied on)
   Manual testing & unit tests from `JetpackMigrationViewModelTest`.

8. What automated tests I added (or what prevented me from doing so)
   Added unit tests to `JetpackMigrationViewModelTest` for the updated language transfer logic.

PR submission checklist:
